### PR TITLE
improv(core): Allow skipping of backreferences

### PR DESCRIPTION
**Pull request by rayzeller**
CAP-8795

Will include Loom in our private repo on the accompanying PR.

This PR allows us to ignore `$` characters in autocomplete regex's. It also addresses a regression when upgrading to 0.1.12.

image

### CODE REVIEW

File: packages/textcomplete-core/src/Dropdown.ts [hunk 1]
```diff
@@ -316,15 +316,15 @@ class DropdownItem {
...
 
     this.el = li
   }
 
   destroy(): this {
     const li = this.el
     li.parentNode?.removeChild(li)
-    li.removeEventListener("click", this.onClick, false)
+    li.removeEventListener("mousedown", this.onClick, false)
     return this
```
Comment (rayzeller) [inline packages/textcomplete-core/src/Dropdown.ts:328]: regression from 0.1.12 that i switched back.

https://github.com/yuku/textcomplete/commit/8d970fca508557e610a6a3468854a6e784dbada8

File: packages/textcomplete-core/src/SearchResult.ts [hunk 2]
```diff
@@ -26,9 +26,12 @@ export class SearchResult<T = unknown> {
...
     if (match == null || match.index == null) return null
-    const replacement = result
-      .replace(MAIN, match[0])
-      .replace(PLACE, (_, p) => match[parseInt(p)])
+    let replacement = result
+      .replace(MAIN, match[0]);
+    
+    if(!this.strategy.skipBackReferences()) {
+      replacement = replacement.replace(PLACE, (_, p) => match[parseInt(p)])
+    }
```
Comment (rayzeller) [inline packages/textcomplete-core/src/SearchResult.ts:34]: Main fix.

**Comment by rayzeller** [comment]
I notated the main code changes. Everything else is distribution code.

File: packages/textcomplete-core/src/Strategy.ts [hunk 3]
```diff
@@ -66,6 +67,10 @@ export class Strategy<T> {
     return this.props.id || null
   }
 
+  skipBackReferences(): boolean | null {
+    return this.props.skipBackReferences || null
+  }
+
```
Comment (rayzeller) [inline packages/textcomplete-core/src/Strategy.ts:73]: Updated props

**Review by kbiedrzycki — COMMENTED** [review]
Do we actually need to commit `dist` files too?

**Comment by rayzeller** [comment]
@kbiedrzycki i'm only about 60% sure actually. there might be a config change that would generate them locally on install. it's not published to npm, so it behaves differently. that being said, I wasn't planning on PR-ing this back to the main repo, given the regression.

think it's worth spending some more time researching?

File: packages/textcomplete-core/src/Strategy.ts [hunk 4]
```diff
@@ -66,6 +67,10 @@ export class Strategy<T> {
     return this.props.id || null
   }
 
+  skipBackReferences(): boolean | null {
+    return this.props.skipBackReferences || null
```
Comment (kbiedrzycki) [inline packages/textcomplete-core/src/Strategy.ts]: Should that have some default (_false_?) value defined somewhere? Shall we allow `null` state for that one?

Comment (rayzeller) [reply packages/textcomplete-core/src/Strategy.ts]: Fixed

Comment (rayzeller) [reply packages/textcomplete-core/src/Strategy.ts]: actually not completely. i see what you're saying. one sec

Comment (rayzeller) [reply packages/textcomplete-core/src/Strategy.ts]: OK it's fixed now

**Review by kbiedrzycki — APPROVED** [review]

**Pull request merged**